Updated CHANGELOG, preparing release 2.2 (#221)
PelionIoT/meta-pelion-edge: CHANGELOG.md

# Changelog

## Pelion Edge 2.2 - January 2021

The primary features in this release:
* [edge-core] Updated Edge Core to [0.15.0](https://github.com/PelionIoT/mbed-edge/blob/master/CHANGELOG.md#release-0150-2021-1-12)

>   * The new FOTA update framework library will be supported on platforms - `imx8mmevk` and `uz3eg-iocc`, (Feb 2021 release) not on `raspberrypi3`. To compile with this library, add following Bitbake parameters to local.conf - `MBED_EDGE_CORE_CONFIG_FIRMWARE_UPDATE="ON"`, `MBED_EDGE_CORE_CONFIG_FOTA_ENABLE="ON"` and `MBED_EDGE_CORE_CONFIG_CURL_DYNAMIC_LINK="ON"`.

Author: The new FOTA update framework is supported on platforms - imx8 and Xilinux board which are expected to release in Feb 2021. RPi do not support that in the coming release.

>   * The new FOTA update framework library will be supported on platforms - `imx8mmevk` and `uz3eg-iocc`, (Feb 2021 release) not on `raspberrypi3`. To compile with this library, add following Bitbake parameters to local.conf - `MBED_EDGE_CORE_CONFIG_FIRMWARE_UPDATE="ON"`, `MBED_EDGE_CORE_CONFIG_FOTA_ENABLE="ON"` and `MBED_EDGE_CORE_CONFIG_CURL_DYNAMIC_LINK="ON"`.

Reviewer: ```suggestion
  * The new FOTA update framework library will be supported on platforms - `imx8mmevk` and `uz3eg-iocc`, but not on `raspberrypi3`. To compile with this library, add following Bitbake parameters to local.conf - `MBED_EDGE_CORE_CONFIG_FIRMWARE_UPDATE="ON"`, `MBED_EDGE_CORE_CONFIG_FOTA_ENABLE="ON"` and `MBED_EDGE_CORE_CONFIG_CURL_DYNAMIC_LINK="ON"`.
```

>   * The new FOTA update framework library will be supported on platforms - `imx8mmevk` and `uz3eg-iocc`, (Feb 2021 release) not on `raspberrypi3`. To compile with this library, add following Bitbake parameters to local.conf - `MBED_EDGE_CORE_CONFIG_FIRMWARE_UPDATE="ON"`, `MBED_EDGE_CORE_CONFIG_FOTA_ENABLE="ON"` and `MBED_EDGE_CORE_CONFIG_CURL_DYNAMIC_LINK="ON"`.

Reviewer: Gotcha. Adding a suggestion since it seems like information is duplicated?
(The header ahready has the release date, unless were going to do another release next month???)

  * The old firmware update library - Update Client (UC) hub is only supported on `raspberrypi3` platform. To enable that add following parameter to local.conf - `MBED_EDGE_CORE_CONFIG_FIRMWARE_UPDATE="ON"`.
* [maestro] Updated maestro to [v2.9.0](https://github.com/PelionIoT/maestro/releases/tag/v2.9.0)

>   * Gateway capabilities - Allows gateways to advertise the features supported by the platform. Maestro utilizes Edge Core's GRM JSON-RPC APIs to add to the gateway device's LwM2M resources. The registered resources are added under Pelion's reserved FeatureMgmt LwM2M object - 33457 with 3 resources - 0 - featureID, 1 - enabled, 2 - config.

Author: Good catch @cmonr. We should link it to this example which was released with the feature - https://github.com/PelionIoT/mbed-edge-examples/blob/master/simple-js-examples/simple-grm-example.js

  * Removed parsing and generation of self-signed certificates. Also removed the platforms rp200 and wwrelay which are no longer supported.
  * Maestro traffic now routed through edge-proxy instead of directly connecting to the Pelion Cloud
* [logrotate] Added bbappend file to install a configuration which rotates the logs of files under `/var/log`.
* [wait-for-pelion-identity] Converted the node program to bash. Removed the generation of self-signed certificate.
* [edge-proxy] Configured services gateway logs (maestro), gateway stats (maestro), relay-term, fluentbit http endpoint to route traffic through edge-proxy and not consume self-signed certificate.
* [meta-nodejs] Removed the dependency on node v8.x. Upgraded the node packages to work with default nodejs version of Dunfell.
* [relay-term] Upgraded relay-term to work with node v12.x. Established an independent recipe from wwrelay-utils and removed the dependency on global-node-modules.

> * [fluentbit] Added recipe to install Fluentbit 1.3.5 on the gateway for providing an open source Log Processor and Forwarder solution.

Reviewer: Should this suggest that it's an option?

Author: Ping @deepakpoornachandra

Author: We decided to disable maestro gateway log by default and only enable Fluentbit.

  * By default, fluentbit is configured with following input endpoints - CPU, MEM, Systemd services - edge-core, edge-proxy, pelion-relay-term, maestro, kubelet, docker and wait-for-pelion-identity and Tail for /wigwag/log/devicejs.log. And an output endpoint HTTP to publish logs at API endpoint - `http://gateways.local:8080/v3/device-logs` (routing through edge-proxy).
* [journald] Enabled Forward Secure Sealing (FSS) feature of systemd journal. This will help detect gateway logs file tampering.

>   * To configure Pelion Edge gateway with sealing key and to keep track of verification key in production setup, use Pelion Edge Provisioner (pep) tool [v2.3.0](https://github.com/PelionIoT/pelion-edge-provisioner/releases/tag/v2.3.0).

Reviewer: More of a personal understanding question.

Sealing key?

Author: The details of that is in Journald FSS document - https://lwn.net/Articles/512895/

>   * Provides the means to make logging persistent. Documentation on how to do this is updated.

Reviewer: Do we need to say where?

Author: ping  @x85446

Author: Updated the Changelog.

* [systemd] Configured to manage the network. By default, disabled Maestro's network management feature.

> * [depreciation] Removed deprecated services.

Reviewer: Making sure, but we're alright with doing this in a minor release, not a major release, right?

Author: Ping @x85446. I think so we are good here because those services were anyways not used in production environment.

  * - devicejs-ng
  * - compatible devicejs-ng protocol translators.
* [image improvements] The "raspberrypi" supported image "console-image" has been simplified and thereby improved.
  * console-image previously, version 1.0 through 2.1, contained Pelion Edge + development tools including but not limited to: compliers, editors, analysis tools, stress tools, and SQA tools.  Pelion Edge version 2.2's console-image contains a minimized set of accompany software for running and testing all of Pelion Edge's software and features.  Important to note, it is not a bare minimal image that strips common Linux tools, but more so what you might expect to find on a heavy embedded device.  With this new strategy, users of the Pelion Edge image can customize the image more easily to their liking.  It is very possible to strip the image more making a more lightweight embedded OS or add more packages to make it more like the image provide previously.
  * meta-pelion-edge itself as a yocto layer is now easier to incorporate with other layers, allowing other yocto projects to incorporate Pelion Edge.
  * [recipe removals]

>     * wwrelay-utils recipe is removed.  Previously this bitbake recipe did many functions which are now replaced.  Reference the recipe additions section below.

Reviewer: Is it necessary to mention that the recipe has been removed if the subsection already mentions that?

Author: ping @x85446 ?

>     * wwrelay-utils recipe is removed.  Previously this bitbake recipe did many functions which are now replaced.  Reference the recipe additions section below.

Reviewer: Also, recipe additions?

Author: @x85446 I think you missed adding the `recipe additions` section.

Reviewer: @x85446 Were you going to add `recipe additions` as well?

### Known issues

- When conducting back-to-back production factory flow with the Pelion Edge Provisioner, the mcc_config directory sometimes is not written correctly and upon reboot, Edge-Core does not connect properly. Workaround: Run the provisioner again.
- Portal is not correctly updated after a firmware campaign in some instances.
- [maestro] The FeatureMgmt config resource is initialized with max 3.8KB of file content. The remaining file content is truncated during initialization. This is most likely due to limitation of gorilla/websocket library, but needs further investigation. Although, Pelion DM users can still push a file size of max 64KB via cloud service APIs.

### Limitations

- The maximum translated devices behind the edge gateway is 100.
- Devices behind Pelion Edge do not support [auto-observation.](https://www.pelion.com/docs/device-management/current/connecting/device-guidelines.html#auto-observation)
- Pelion Device Management Client enabled devices must first boostrap to the Pelion Device Management Cloud before connecting to Pelion Edge.
- No moving devices are supported. (Device would be moving from Edge to another Edge device.)

### Important note

> While provisioning your gateway, please use `vendor-id=42fa7b48-1a65-43aa-890f-8c704daade54` to unlock the rich node features, such as gateway logs and gateway terminal in the Pelion web portal.

Reviewer 2: Please mention that fota flag needs to be enabled while creating the build for this feature.

Author: @udayakumar17 its mentioned here - https://github.com/PelionIoT/meta-pelion-edge/pull/221/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR8 
We will also add that to the public doc, brought in by - https://github.com/PelionIoT/pelion-dm-edge-docs/pull/108

* [logrotate] Removed crontab method to call logrotate every 5 mins. Instead configured systemd logrotate timer to 5 mins.
* [identity-tool] Established an independent recipe under recipes-connectivity and removed the dependency on wwrelay-utils.
  * Converted the identity node program to bash which now lives in a new project called pe-utils.
  * Installed v2.0.4 of identity-tool, which majorly removes the generation of self-signed certificate.
* [edge-proxy] Configured services gateway logs (maestro), gateway stats (maestro), relay-term, fluentbit http endpoint to route traffic through edge-proxy and not consume self-signed certificate.
* [meta-nodejs] Removed the dependency on node v8.x. Upgraded the node packages to work with default nodejs version of Dunfell.
* [relay-term] Upgraded relay-term to work with node v12.x. Established an independent recipe from wwrelay-utils and removed the dependency on global-node-modules.
* [mbed-fcce] Upgraded factory-configurator-client-example to v4.7.0 and pinned its respective dependencies.
* [fluentbit] Added recipe to install Fluentbit 1.3.5 on the gateway for providing an open source Log Processor and Forwarder solution.

>   * By default, fluentbit is configured with following input endpoints - CPU, MEM, Systemd services - edge-core, edge-proxy, pelion-relay-term, maestro, kubelet, docker and wait-for-pelion-identity. And an output endpoint HTTP to publish logs at API endpoint - `http://gateways.local:8080/v3/device-logs` (routing through edge-proxy).

Reviewer 3: Could you please rephrase this bullet? I'm not sure what we're trying to say.

Author: That is a bit technical. Someone who is familiar with Fluent Bit will get that. Should we simplify it?

# Changelog

> ## Pelion Edge 2.2 - January 2021

Reviewer: ```suggestion
## Pelion Edge 2.2 - February 2021
```

Not sure if this needs to be changed now.

Author: Sure. Added this to the next commit.

> ## Pelion Edge 2.2 - January 2021

Reviewer: Adding a different note. @ygoyal18 this file should include the change for the hotifx that was needed a week or two ago: https://github.com/PelionIoT/meta-pelion-edge/commit/a033908508908dae82d0c746e958aa3e9f44723b

Author: Sure. Will add that.

The primary features in this release:

* [edge-core] Updated Edge Core to [0.15.0](https://github.com/PelionIoT/mbed-edge/blob/master/CHANGELOG.md#release-0150-2021-1-12).
  * Moved the mbed-edge-core recipe to its own meta layer - [meta-mbed-edge](https://github.com/PelionIoT/meta-mbed-edge).

>   * The new FOTA update framework library is supported on platforms `imx8mmevk` and `uz3eg-iocc` but not on `raspberrypi3`. To compile with this library, add the `MBED_EDGE_CORE_CONFIG_FIRMWARE_UPDATE="ON"`, `MBED_EDGE_CORE_CONFIG_FOTA_ENABLE="ON"` and `MBED_EDGE_CORE_CONFIG_CURL_DYNAMIC_LINK="ON"` Bitbake parameters to local.conf.

Reviewer: This needs to be sync'd with @esajaa.

Afte running multiple tests, I've found that LmP sources the environment and variables in a different way than Yocto. I also think that the job that @OPpuolitaival and @esajaa have been working on add the variables in a different location.

Reviewer 4: What's the difference? Both yocto and lmp builds add the variables to the local.conf? Or so I understood from Yocto quick start  https://developer.pelion.com/docs/device-management-edge/2.1/quick-start/yocto-quick-start.html

Reviewer: @esajaa Unless something's changed within the last day, the way LmP sources its variables is a bit more involved than Yocto, and a bit strange.

Should it be mentioned that the location for LmP's 'local.conf' resides within a hidden directory?

I've also had mixed experiences with setting your own variables within `.repo/manifests/conf/local.conf`, mainly because it seems that when LmP sources `setup-environment` for the first time, it's doing a lot more than what a Yocto env would do.

I've been testing it today, but have had mixed results, but definitely know that a local.conf file is not available prior to invoking the `source setup-environemnt build-factory` command.

Reviewer 4: Ah, I think I see where the confusion comes from. In theory there are now two local.conf.samples. One comes from meta-pelion-edge, and one comes from repo containing lmp-manifest (since lmp uses the local.conf a bit different way). LmP doesn't utilize the local.conf.sample from meta-pelion-edge, so any changes there won't be applied. Possible new variables have to be added to this .repo folder local.conf.sample.

local.conf file should not be present before running setup-environment in any type build. What setup-environment does is that it pulls some local.conf.sample, and places it to the build folder with name local.conf. It may be that lmp setup-environment does some extra things with variables, but that's at least the behavior with local.conf

After the local.conf is set, running the setup-environment doesn't pull the local.conf.sample anymore

Reviewer: > local.conf file should not be present before running setup-environment in any type build. What setup-environment does is that it pulls some local.conf.sample, and places it to the build folder with name local.conf

Yup. It's this nuanced detail that I think should be present in some way. Otherwise anyone thinking that they can set things using the normal local.conf method will have issues.

Reviewer: Resolved with b016b00867efe85d73d6deb1808f89e48b2ec337

>   * The old firmware update library, Update Client (UC) hub, is only supported on the `raspberrypi3` platform. To enable that, add the `MBED_EDGE_CORE_CONFIG_FIRMWARE_UPDATE="ON"` parameter to local.conf.

Reviewer: Is this for both Yocto and LmP?

Reviewer 4: LmP doesn't support Raspberrypi atm

Reviewer: Should the comment be updated to add that detail?

* [maestro] Updated Maestro to [v2.9.0](https://github.com/PelionIoT/maestro/releases/tag/v2.9.0).

>   * Gateway capabilities - Allows gateways to advertise the features supported by the platform. Maestro uses Edge Core's GRM JSON-RPC APIs to add to the gateway device's LwM2M resources. The registered resources are added under Pelion's reserved FeatureMgmt LwM2M object - 33457 with three resources - 0 - featureID, 1 - enabled, 2 - config.

Reviewer: I think I recall adding a comment about these same numbers somewhere, but I don't recall if it was resolved.

* [relay-term] Upgraded relay-term to work with node v12.x. Established an independent recipe from wwrelay-utils and removed the dependency on global-node-modules.
* [mbed-fcce] Upgraded factory-configurator-client-example to v4.7.0 and pinned its dependencies.
* [fluentbit] Added recipe to install Fluentbit 1.3.5 on the gateway for providing an open source log processor and forwarder solution.
  * By default, Fluent Bit is configured with following input endpoints - CPU, MEM, Systemd services - edge-core, edge-proxy, pelion-relay-term, maestro, kubelet, docker and wait-for-pelion-identity. And an output endpoint HTTP to publish logs at API endpoint - `http://gateways.local:8080/v3/device-logs` (routing through edge-proxy).
* [journald] Enabled Forward Secure Sealing (FSS) feature of systemd journal. This helps detect gateway logs file tampering.
  * To configure Pelion Edge gateway with a sealing key and to keep track of the verification key in production setup, use Pelion Edge Provisioner (PEP) tool [v2.3.0](https://github.com/PelionIoT/pelion-edge-provisioner/releases/tag/v2.3.0).

>   * By default, the gateway is configured with persistent journal logging. To disable that, set flag `VOLATILE_LOG_DIR = "no"` in `local.conf` and also update the `Storage` in recipes-core/systemd/systemd-conf/journald.conf. Note: If you disable persistent logging, the FSS feature won't work.

Reviewer: Yocto and LmP?

Reviewer 5: I have a new section at the top of the Changelog that specifies the only Pelion Edge feature difference.  That should address the question as to which of these bullets apply to which OS.

* [systemd] Configured to manage the network. By default, we disabled Maestro's network management feature.
* [systemd] Updated the unit startup sequence and changed `Requires` directive to `Wants`. This makes the system more robust when dealing with failing services.

  ```
  network-online --> edge-core --> wait-for-pelion-identity --> edge-proxy
                                                            --> edge-kubelet
                                                            --> fluentbit (td-agent-bit)
                                                            --> pelion-relay-term
                                                            --> maestro --> devicedb
  ```
  
* [deprecation] Removed deprecated services.
   * devicejs-ng.
   * compatible devicejs-ng protocol translators.
* [image improvements] The "raspberrypi" supported image "console-image" has been simplified and improved.

>    * console-image previously, version 1.0 through 2.1, contained Pelion Edge + development tools including but not limited to: compliers, editors, analysis tools, stress tools, and SQA tools.  Pelion Edge version 2.2's console-image contains a minimized set of accompany software for running and testing all of Pelion Edge's software and features.  Important to note, it is not a bare minimal image that strips common Linux tools, but more so what you might expect to find on a heavy embedded device.  With this new strategy, users of the Pelion Edge image can customize the image more easily to their liking.  It is very possible to strip the image more making a more lightweight embedded OS or add more packages to make it more like the image provide previously.

Reviewer: Is it already understood that `console-iamge` only referrs to the Yocto build?

Reviewer 5: added a subbullet for LMP

  * Removed parsing and generation of self-signed certificates. Also removed the platforms rp200 and wwrelay which are no longer supported.
  * Maestro traffic now routed through edge-proxy instead of directly connecting to the Pelion Cloud
* [logrotate] Added bbappend file to install a configuration which rotates the logs of files under `/var/log`.
* [wait-for-pelion-identity] Converted the node program to bash. Removed the generation of self-signed certificate.
* [edge-proxy] Configured services gateway logs (maestro), gateway stats (maestro), relay-term, fluentbit http endpoint to route traffic through edge-proxy and not consume self-signed certificate.
* [meta-nodejs] Removed the dependency on node v8.x. Upgraded the node packages to work with default nodejs version of Dunfell.
* [relay-term] Upgraded relay-term to work with node v12.x. Established an independent recipe from wwrelay-utils and removed the dependency on global-node-modules.
* [fluentbit] Added recipe to install Fluentbit 1.3.5 on the gateway for providing an open source Log Processor and Forwarder solution.

>   * By default, fluentbit is configured with following input endpoints - CPU, MEM, Systemd services - edge-core, edge-proxy, pelion-relay-term, maestro, kubelet, docker and wait-for-pelion-identity and Tail for /wigwag/log/devicejs.log. And an output endpoint HTTP to publish logs at API endpoint - `http://gateways.local:8080/v3/devicejs-logs` (routing through edge-proxy).

Reviewer 6: ```suggestion
  * By default, fluentbit is configured with following input endpoints - CPU, MEM, Systemd services - edge-core, edge-proxy, pelion-relay-term, maestro, kubelet, docker and wait-for-pelion-identity and Tail for /wigwag/log/devicejs.log. And an output endpoint HTTP to publish logs at API endpoint - `http://gateways.local:8080/v3/device-logs` (routing through edge-proxy).
```

Reviewer 6: Minor change.
